atomic_reactor/utils/resource_manager.py

```diff
@@ -67,186 +166,378 @@ def username(self):
+            raise SlotWriteError(_errmsg)
 
-                cmd = self.get_cmd_to_write_empty_slot(slot_id)
-                _, _, exitcode = self._ssh_run_remote_cmd(client, cmd)
+    @property
+    def is_valid(self):
+        """ Check whether the content is valid """
+        return self._data.is_valid
 
-                if exitcode != 0:
-                    logger.warning("Unlocking slot %s on host %s failed!", slot_id, self.hostname)
-            else:
-                logger.warning("%s - slot %s is available. Unlocking skipped!",
-                               self.hostname, slot_id)
+    @property
+    def is_free(self) -> bool:
+        """ Check whether the slot is in free state """
+        return self._data.is_empty
+
+    def is_locked_by(self, prid: str) -> bool:
+        """ Check whether the slot is locked by a pipelinerun """
+        return self._data.prid == prid
+
+    def lock(self, prid: str) -> bool:
+        """ Lock the slot for a pipelinerun """
+        if not self.is_free:
+            logger.debug("%s: slot %s is not free, unable to lock it",
+                         self.hostname, self.id)
+            return False
+
+        if not self.is_valid:
+            logger.warning("%s: slot %s contains invalid content, it's probably corrupted, "
+                           "unable to lock it.", self.hostname, self.id)
+            return False
+
+        data = SlotData(prid=prid, timestamp=datetime.utcnow().isoformat())
+        self._write(data.to_string())
+        return True
+
+    def unlock(self, prid: str) -> bool:
+        """ Unlock the slot for a pipelinerun """
+        if self.is_free:
+            logger.warning("%s: slot %s is free, skip unlocking", self.hostname, self.id)
+            # Should we return False instead?
+            return True
+
+        if not self.is_valid:
+            logger.warning("%s: slot %s contains invalid content, it's probably corrupted, "
+                           "unable to unlock it.", self.hostname, self.id)
+            return False
+
+        if not self.is_locked_by(prid):
+            logger.warning("%s: cannot unlock slot %s, it's not locked by %s",
+                           self.hostname, self.id, prid)
+            return False
+
+        # Empty the slot
+        self._write()
+        return True
 
 
 class LockedResource:
 
     def __init__(self, host: RemoteHost, slot: int, prid: str):
+        """ Instantiate a locked resource with remote host, slot id and pipelinerun id
+
+        :param host: RemoteHost, RemoteHost instance
+        :param slot: int, slot ID
+        :param prid: str, pipeline run ID
         """
-        :param slot: int, Remote host slot ID
-        :param prid: str, Pipeline run ID
-        """
+        self.host = host
         self.slot = slot
         self.prid = prid
-        self.host = host
 
     def unlock(self):
+        """ Unlock the resource for pipelinerun """
         self.host.unlock(self.slot, self.prid)
```

Comment:
This ignores the return value of `host.unlock()`.

Which brings me to the next point - since `LockedResource.unlock()` and `RemoteHost.lock()|unlock()` are part of the public API of this module, maybe returning booleans to indicate failure is not the best way to go. It could be easy to forget to check the return value and assume that calling `unlock()` (and not getting an exception) means the slot is unlocked, kind of like it happened here.

Reply:
This concern makes sense, the problem is we can't do anything when `unlock` failed, so I just ignore it silently, there could be the following cases that `unlock` fails:
1. slot is not owned by this pr, in theory, it should never happen.
2. slot is not locked, in theory it should never happen.
3. slot operations failed to due connection errors.

when I say "it should never happen", it doesn't count the situations of system reboots, slot be changed by people manually.
For case 1, we should just ignore that, and let clean up jobs to fix bad slots.
For case 2, we can also ignore the return value, the free slot will be used by other pipelinerun.
For case 3, maybe we should add retry policy to `host.lock` and `host.unlock` to handle it, and can still ignore the returned value, wdyt?

Reply:
For case 3, that should already be handled by the `SSHRetrySession`, right? I think the behavior is fine, my only concern is that we are using booleans to indicate the success/failure of an operation which is normally expected to succeed - that is what exceptions are for.

But it's not a dealbreaker for me. Thinking about the integration with the build code, we will only really use `RemoteHostPool.lock_resource()` and `LockedResource.unlock()` anyway, and those pretty much behave as expected. Like you said, there's not much we can do with `unlock()` failures except ignore them, anyway.

Reply:
> For case 3, that should already be handled by the SSHRetrySession, right?

yes, I should say command errors or so on instead, which won't be captured by SSHRetrySession, they raise `SlotLockError`, `SlotReadError` or `SlotWriteError`.  I've updated patch to retry on these exceptions.